lib/FrontendTool/FrontendTool.cpp

```diff
@@ -932,11 +947,12 @@ static bool performCompile(CompilerInstance &Instance,
   // We've just been told to perform a typecheck, so we can return now.
   if (Action == FrontendOptions::ActionType::Typecheck) {
     const bool hadPrintAsObjCError = printAsObjCIfNeeded(
-        opts.InputsAndOutputs.supplementaryOutputs().ObjCHeaderOutputPath,
+        Invocation.getObjCHeaderOutputPathForAtMostOnePrimary(),
         Instance.getMainModule(), opts.ImplicitObjCHeaderPath, moduleIsPublic);
 
-    const bool hadEmitIndexDataError = emitIndexDataIfNeeded(
-        Instance.getPrimarySourceFile(), Invocation, Instance);
+    const bool hadEmitIndexDataError =
+        emitIndexDataForAllPrimariesOrModuleIfNeeded(Invocation, Instance);
```

Comment:
Nitpick: does this really need a separate name? `emitIndexDataIfNeeded` that doesn't take a particular file seems pretty clear.

Reply:
Many of these functions only deal with a single primary. Why not use a longer name for the sake of clarity? When I made these changes, I had to inspect functions like `emitIndexData` and follow the call chains down to see what they did. A doxygen comment could give the info, but that takes an additional gesture to pull it up. Or, one can gather that a specific primary isn't passed as an argument, if one understands that Instance is primary-invarient (which a noob might not),
but one cannot gather that non-primaries do not have index data emitted. But I'll change it back if you like.

Reply:
I was definitely basing this on the arguments. `emitFoo(SF, Instance)` is almost certainly specific to `SF`, and therefore `emitFoo(Instance)` would *not* be in a world where batch mode is commonplace. The exact meaning of "not just one file" is ambiguous, I suppose—is it all primaries or all inputs?—but in practice we don't do *anything* for non-primary files.

For me, I'd balance this by putting the explanation in a comment, but not putting it in the function name. It's adding more cognitive burden than it's reducing.

Reply:
I've changed it to `emitIndexData` with a Doxygen comment.